### PR TITLE
Create Pod+PodStatus resource lifecycle test - +4 endpoint coverage

### LINKED ISSUE #90936: Write CoreV1Pod and CoreV1PodStatus lifecycle test - +4 coverage
# Progress `[7/8]`
* [x] Original approval issue: #88545
* [x]  Test approval issue: [kubernetes/kubernetes#90936](https://github.com/kubernetes/kubernetes/issues/90936)
* [x]  [APISnoop PR](https://github.com/cncf/apisnoop/pull/290)
* [x]  Test pr: [PR](https://github.com/kubernetes/kubernetes/pull/90942) - Reverted due to flakes
* [x] New Test [PR](https://github.com/kubernetes/kubernetes/pull/93459)
* [x]  Two weeks soak start date: 7 November 2020 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&include-filter-by-regex=should%20run%20through%20the%20lifecycle%20of%20Pods%20and%20PodStatus) 
* [ ]  Two weeks soak end date: 20 November 2020
* [x] Test promotion pr: [#96485](https://github.com/kubernetes/kubernetes/pull/96485)

This issue is created to allow edit by @Riaankl

/wip
/hold

# Identify an untested feature Using APISnoop
According to this APIsnoop query, there are still some remaining Pod endpoints which are untested.

```
SELECT
  operation_id,
  -- k8s_action,
  -- path,
  -- description,
  kind
  FROM untested_stable_core_endpoints
  where path not like '%volume%'
  and kind like 'Pod'
  -- and kind like ''
  -- and operation_id ilike '%%'
 ORDER BY kind,operation_id desc
 -- LIMIT 25
       ;
```

```
            operation_id             | kind 
-------------------------------------|------
 replaceCoreV1NamespacedPodStatus    | Pod
 readCoreV1NamespacedPodStatus       | Pod
 listCoreV1PodForAllNamespaces       | Pod
 deleteCoreV1CollectionNamespacedPod | Pod
(4 rows)
```

# Use API Reference to Lightly Document the Feature
* [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
* [client-go - Pod](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/pod.go)

# The mock test
## Test outline
1. Create a Pod with a static label
2. Patch the Pod with a new Label and updated data
3. Get the Pod to ensure it's patched
4. Read the Pod's status
5. Replace the Pod's status Ready condition to False
6. Read the Pod's status to check if Ready is False
7. List all Pods in all Namespaces find the Pod(1) ensure that the Pod is found and is patched
8. Delete Namespaced Pod(1) via a Collection with a LabelSelector
9. Get the Pod to check that it's deleted

## Example in Go
```go
package main
import (
  "encoding/json"
  "time"
  "fmt"
  "flag"
  "os"
  v1 "k8s.io/api/core/v1"
  "k8s.io/client-go/dynamic"
  "k8s.io/apimachinery/pkg/runtime/schema"
  metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
  "k8s.io/client-go/kubernetes"
  "k8s.io/apimachinery/pkg/types"
  "k8s.io/client-go/tools/clientcmd"
)
func main() {
  // uses the current context in kubeconfig
  kubeconfig := flag.String("kubeconfig", fmt.Sprintf("%v/%v/%v", os.Getenv("HOME"), ".kube", "config"), "(optional) absolute path to the kubeconfig file")
  flag.Parse()
  config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
  if err != nil {
      fmt.Println(err)
      return
  }
  // make our work easier to find in the audit_event queries
  config.UserAgent = "live-test-writing"
  // creates the clientset
  ClientSet, _ := kubernetes.NewForConfig(config)
  DynamicClientSet, _ := dynamic.NewForConfig(config)
  podResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
  testNs := "default"
  testPodName := "pod-test"
  testPodImage := "nginx"
  testPodImage2 := "nginx"
  testPod := v1.Pod{
      ObjectMeta: metav1.ObjectMeta{
          Name: testPodName,
          Labels: map[string]string{
              "test-pod-static": "true",
          },
      },
      Spec: v1.PodSpec{
          Containers: []v1.Container{
              {
                 Name: testPodName,
                 Image: testPodImage,
              },
          },
      },
  }
  // create Pod with a static label
  _, err = ClientSet.CoreV1().Pods(testNs).Create(&testPod)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] created Pod")
  // setup a watch for the RC
  podWatch, err := ClientSet.CoreV1().Pods(testNs).Watch(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  podWatchChan := podWatch.ResultChan()
  fmt.Println("[status] watching for Pod to be ready")
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
         fmt.Println("Unable to fix type")
         return
      }
      if podEvent.Status.Phase == "Running" {
         break
      }
  }
  fmt.Println("[status] Pod is Ready")
  // patch the Pod with a new Label and updated data
  podPatch, err := json.Marshal(map[string]interface{}{
      "metadata": map[string]interface{}{
          "labels": map[string]string{
              "podtemplate": "patched",
          },
      },
      "spec": map[string]interface{}{
          "containers": []map[string]interface{}{
              {
                  "name": testPodName,
                  "image": testPodImage2,
              },
          },
      },
  })
  if err != nil {
      fmt.Println(err)
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).Patch(testPodName, types.StrategicMergePatchType, []byte(podPatch))
  if err != nil {
      fmt.Println(err)
      return
  }
  // get the Pod and ensure it's patched
  pod, err := ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err != nil {
      fmt.Println(err)
      return
  }
  if pod.ObjectMeta.Labels["test-pod-static"] != "true" || pod.Spec.Containers[0].Image != testPodImage2 {
      fmt.Println("[error] patching of Pod failed")
      return
  }
  // get pod status
  podStatusUnstructured, err := DynamicClientSet.Resource(podResource).Namespace(testNs).Get(testPodName, metav1.GetOptions{}, "status")
  if err != nil {
      fmt.Println(err)
      return
  }
  podStatusUjson, _ := json.Marshal(podStatusUnstructured)
  var podStatus v1.Pod
  json.Unmarshal(podStatusUjson, &podStatus)
  // replace the Pod's status Ready condition to False
  podStatusUpdated := podStatus
  podStatusFieldPatchCount := 0
  podStatusFieldPatchCountTotal := 2
  for pos, cond := range podStatusUpdated.Status.Conditions {
      if (cond.Type == "Ready" && cond.Status == "True") || (cond.Type == "ContainersReady" && cond.Status == "True") {
          podStatusUpdated.Status.Conditions[pos].Status = "False"
          podStatusFieldPatchCount++
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Println("[error] failed to patch all relevant Pod conditions")
      return
  }
  _, err = ClientSet.CoreV1().Pods(testNs).UpdateStatus(&podStatusUpdated)
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] updated PodStatus")
  // list all Pods and get their status to ensure it's Ready condition is False
  podsList, err := ClientSet.CoreV1().Pods("").List(metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] fetched all Pods by LabelSelector")
  podStatusFieldPatchCount = 0
  podStatusFieldPatchCountTotal = 2
  for _, podItem := range podsList.Items {
      for _, cond := range podItem.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
  }
  if podStatusFieldPatchCount != podStatusFieldPatchCountTotal {
      fmt.Printf("[error] failed to update PodStatus - %v/%v conditions failed to update (%v, %v)", podStatusFieldPatchCount, podStatusFieldPatchCountTotal, "Ready", "ContainersReady")
      return
  }
  fmt.Println("[status] PodStatus was updated successful")
  // delete the Pod via a Collection with a LabelSelector
  err = ClientSet.CoreV1().Pods(testNs).DeleteCollection(&metav1.DeleteOptions{}, metav1.ListOptions{LabelSelector: "test-pod-static=true"})
  if err != nil {
      fmt.Println(err)
      return
  }
  fmt.Println("[status] watching for Pod to be not Ready")
  podEventChannel:
  for event := range podWatchChan {
      podEvent, ok := event.Object.(*v1.Pod)
      if ok != true {
          fmt.Println("Unable to fix type")
          return
      }
      podStatusFieldPatchCount := 0
      podStatusFieldPatchCountTotal := 2
      for _, cond := range podEvent.Status.Conditions {
          if (cond.Type == "Ready" && cond.Status == "False") || (cond.Type == "ContainersReady" && cond.Status == "False") {
              podStatusFieldPatchCount++
          }
      }
      if podStatusFieldPatchCount == podStatusFieldPatchCountTotal {
          break podEventChannel
      }
  }
  time.Sleep(5 * time.Second)
  fmt.Println("[status] Pod no longer available")
  // fetch the Pod to check if it's deleted
  _, err = ClientSet.CoreV1().Pods(testNs).Get(testPodName, metav1.GetOptions{})
  if err == nil {
      fmt.Println("[error] Pod still available after deletion; failed to delete Pod")
      return
  }
  // write test here
  fmt.Println("[status] complete")
}
```

```go
[status] created Pod
[status] watching for Pod to be ready
[status] Pod is Ready
[status] updated PodStatus
[status] fetched all Pods by LabelSelector
[status] PodStatus was updated successful
[status] watching for Pod to be not Ready
[status] Pod no longer available
[status] complete
```

# Verify Increase it Coverage with APISnoop
Discover useragents:

```
select distinct useragent from audit_event where bucket='apisnoop' and useragent not like 'kube%' and useragent not like 'coredns%' and useragent not like 'kindnetd%' and useragent like 'live%';
```

```
     useragent     
-------------------
 live-test-writing
(1 row)
```

List endpoints hit by the test:

```
select * from endpoints_hit_by_new_test where useragent like 'live%'; 
```

```
     useragent     |            operation_id             | hit_by_ete | hit_by_new_test 
-------------------|-------------------------------------|------------|-----------------
 live-test-writing | readCoreV1NamespacedPodStatus       | f          |               1
 live-test-writing | readCoreV1NamespacedPod             | t          |               2
 live-test-writing | deleteCoreV1CollectionNamespacedPod | f          |               2
 live-test-writing | patchCoreV1NamespacedPod            | t          |               2
 live-test-writing | listCoreV1PodForAllNamespaces       | f          |               1
 live-test-writing | createCoreV1NamespacedPod           | t          |               2
 live-test-writing | replaceCoreV1NamespacedPodStatus    | f          |               2
 live-test-writing | listCoreV1NamespacedPod             | t          |               1
(8 rows)
```

Display endpoint coverage change:

```
select * from projected_change_in_coverage;
```

```
   category    | total_endpoints | old_coverage | new_coverage | change_in_number 
---------------|-----------------|--------------|--------------|------------------
 test_coverage |             438 |          190 |          194 |                4
(1 row)
```

# Final notes
If a test with these calls gets merged, ****test coverage will go up by 4 points****

This test is also created with the goal of conformance promotion.

/sig testing
/sig architecture
/area conformance

### LINKED ISSUE #90957: Watch event test verification tooling
* [Ticket status](#sec-1)
* [Current problem](#sec-2)
* [Proposed solution](#sec-3)

This Issue replace #90574

# Ticket status
* [ ]  [K8s watch events test tooling #323](https://github.com/cncf/apisnoop/pull/323)
* [ ]  [Watch event test verification tooling #90574](https://github.com/kubernetes/kubernetes/issues/90574)
* [X] [PR #91416 ](https://github.com/kubernetes/kubernetes/pull/91416)
# Current problem
Watch events may be missed and the test may still pass. We need to ensure that all expected watch events are seen.

Watch events are not collected and checked at the end of the test.

# Proposed solution
In order to solve the problem, we must collect all watch events as they come in and perform a check at the end of the test to ensure that the order is correct and they contain no errors.

```go
package main

import (
  "fmt"
)

func main() {
  expectedWatchEventsAll := [][]string{
    {
      "ADDED",
      "ADDED",
      "DELETED",
    },
    {
      "ADDED",
      "MODIFIED",
      "MODIFIED",
      "DELETED",
    },
    {
      "ADDED",
      "MODIFIED",
      "DELETED",
    },
  }
  for expectedExampleInt, expectedWatchEvents := range expectedWatchEventsAll {
	  fmt.Println("Running example", expectedExampleInt)
    allowedAttempts := 3
    failure := VerifyWatchEventOrder(allowedAttempts, expectedWatchEvents, func() []string {
      var myWatchEvent string
      var watchEvent []string
      fmt.Println("An event takes place (1/3)")
      myWatchEvent = "ADDED"
      watchEvent = append(watchEvent, myWatchEvent)

      fmt.Println("An event takes place (2/3)")
      myWatchEvent = "MODIFIED"
      watchEvent = append(watchEvent, myWatchEvent)

      fmt.Println("An event takes place (3/3)")
      myWatchEvent = "DELETED"
      watchEvent = append(watchEvent, myWatchEvent)

      return watchEvent
    })
    if failure != "" {
      fmt.Println(failure, "watch events occured in the wrong or incorrect order")
    }
    fmt.Println("Check complete")
  }
}

func VerifyWatchEventOrder(retries int, expectedWatchEvents []string, scenario func() []string) (failure string) {
  var attemptContainsFailure bool
attempts:
  for try := 1; try <= retries; try++ {
    if try == retries {
      return failure
    }
    actualWatchEvents := scenario()
    if len(expectedWatchEvents) != len(actualWatchEvents) {
      failure = "expected watch events count does not match the count of actual watch events"
      attemptContainsFailure = true
      continue
    }
    for watchEventInt, watchEvent := range actualWatchEvents {
      if expectedWatchEvents[watchEventInt] != watchEvent {
        failure = fmt.Sprintf("(index %v) %v not found, found %v instead. %v/%v attempts", watchEventInt, expectedWatchEvents[watchEventInt], watchEvent, try, retries)
        attemptContainsFailure = true
      }
    }
    if attemptContainsFailure == false {
      failure = ""
      break attempts
    }
  }
  return failure
}
```

```
Running example 0
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
(index 1) ADDED not found, found MODIFIED instead. 2/3 attempts watch events occured in the wrong or incorrect order
Check complete
Running example 1
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
expected watch events count does not match the count of actual watch events watch events occured in the wrong or incorrect order
Check complete
Running example 2
An event takes place (1/3)
An event takes place (2/3)
An event takes place (3/3)
Check complete
```

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This PR adds a test to test the following untested endpoints:

* replaceCoreV1NamespacedPodStatus
* readCoreV1NamespacedPodStatus
* listCoreV1PodForAllNamespaces
* deleteCoreV1CollectionNamespacedPod

**Which issue(s) this PR fixes**:
Fixes #90936
Resolves #90957 

**Special notes for your reviewer**:
Adds +4 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?**:

```
NONE
```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```
NONE
```

/sig testing

### CODE REVIEW

**Comment by Riaankl** [comment]
https://github.com/kubernetes/kubernetes/pull/89453#issuecomment-624397465
This PR replace #89453

**Comment by hh** [comment]
From [comment from @spiffxp](https://github.com/kubernetes/kubernetes/pull/89453#issuecomment-618606579

> This PR has been extraordinarily unlucky with the flakes it has hit across jobs, but it looks like it's going to introduce a new one. Can you please address?

**Comment by Riaankl** [comment]
pull-kubernetes-e2e-gce-ubuntu-containerd

**Comment by Riaankl** [comment]
pull-kubernetes-e2e-gce-device-plugin-gpu

**Comment by Riaankl** [comment]
/assign @spiffxp @johnbelamaric

@BobyMCbobs update this test and removed Watch tooling
Can we please get an /LGTM, /approve and /milestone 1.19

File: test/e2e/common/pods.go [hunk 1]
```diff
@@ -870,6 +875,119 @@ var _ = framework.KubeDescribe("Pods", func() {
...
+			},
+		})
+		framework.ExpectNoError(err, "failed to marshal JSON patch for Pod")
+		_, err = f.ClientSet.CoreV1().Pods(testNs).Patch(context.TODO(), testPodName, types.StrategicMergePatchType, []byte(podPatch), metav1.PatchOptions{})
+		framework.ExpectNoError(err, "failed to patch Pod %s in namespace %s", testPodName, testNs)
+
+		ginkgo.By("getting the Pod and ensuring that it's patched")
+		pod, err := f.ClientSet.CoreV1().Pods(testNs).Get(context.TODO(), testPodName, metav1.GetOptions{})
+		framework.ExpectNoError(err, "failed to fetch Pod %s in namespace %s", testPodName, testNs)
+		framework.ExpectEqual(pod.ObjectMeta.Labels["test-pod-static"], "true", "failed to patch Pod - missing label")
```
Comment (spiffxp) [inline test/e2e/common/pods.go:941]: This is verifying that the old label still exists (which is a valid check to do) but it's unclear to me whether you intended to check that the new label also exists?

File: test/e2e/common/pods.go [hunk 2]
```diff
@@ -870,6 +875,119 @@ var _ = framework.KubeDescribe("Pods", func() {
...
+
+		ginkgo.By("check the Pod again to ensure its Ready conditions are False")
+		podStatusFieldPatchCount = 0
+		podStatusFieldPatchCountTotal = 2
+		for _, cond := range podStatusUpdate.Status.Conditions {
+			if (cond.Type == v1.PodReady && cond.Status == v1.ConditionFalse) || (cond.Type == v1.ContainersReady && cond.Status == v1.ConditionFalse) {
+				podStatusFieldPatchCount++
+			}
+		}
+		framework.ExpectEqual(podStatusFieldPatchCount, podStatusFieldPatchCountTotal, "failed to update PodStatus - field patch count doesn't match the total")
```
Comment (spiffxp) [inline test/e2e/common/pods.go:975]: I think you're going to be in a race with the kubelet to see whose updates to the pod's conditions are going to be present when it comes time to read the conditions again.

Have you considered patching a condition the kubelet doesn't normally write to?

File: test/e2e/common/pods.go [hunk 3]
```diff
@@ -870,6 +875,119 @@ var _ = framework.KubeDescribe("Pods", func() {
...
+		podWatchChan := podWatch.ResultChan()
+		ginkgo.By("watching for Pod to be ready")
+		eventFound := false
+		for watchEvent := range podWatchChan {
+			if podWatchEvent, ok := watchEvent.Object.(*v1.Pod); ok && podWatchEvent.Status.Phase == v1.PodRunning {
+				eventFound = true
+				break
+			}
+		}
+		framework.ExpectEqual(eventFound, true, "failed to find running Pod %s in namespace %s", testPodName, testNs)
```
Comment (spiffxp) [inline test/e2e/common/pods.go:920]: This watch loop is going to be susceptible to disconnects, have you considered using `watchtools.Until`?

**Review by spiffxp — CHANGES_REQUESTED** [review]
I strongly suspect this is going to flake.  Find some part of Pod Status that wouldn't normally be overwritten by kubelet, or prove to me the condition fields you've chosen don't cause flakes under normal e2e load.

File: test/e2e/common/pods.go (hunk 1 above)
Comment (BobyMCbobs) [reply test/e2e/common/pods.go]: I agree, so in bd34e1c445b, I've updated it to patch and check labels that make sense

File: test/e2e/common/pods.go (hunk 3 above)
Comment (BobyMCbobs) [reply test/e2e/common/pods.go]: Updated in d60a57ad738

File: test/e2e/common/pods.go (hunk 2 above)
Comment (BobyMCbobs) [reply test/e2e/common/pods.go]: Perhaps something like _startTime_?

https://github.com/kubernetes/api/blob/master/core/v1/types.go#L3618

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

File: test/e2e/common/pods.go [hunk 4]
```diff
@@ -870,6 +875,154 @@ var _ = framework.KubeDescribe("Pods", func() {
 		err = wait.PollImmediate(podRetryPeriod, podRetryTimeout, checkPodListQuantity(f, "type=Testing", 0))
 		framework.ExpectNoError(err, "found a pod(s)")
 	})
+
+	ginkgo.It("should run through the lifecycle of Pods and PodStatus", func() {
```
Comment (liggitt) [inline test/e2e/common/pods.go:879]: This test is flaking. See https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/93398/pull-kubernetes-e2e-gce-ubuntu-containerd/1286493826562658304/ as an example

**Comment by liggitt** [comment]
See failures in 
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-ubuntu-containerd/1286493826562658304
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce-ubuntu-containerd/1286474574245924867
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286474574245924868
* https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-kind/1286412665119313926

revert opened at https://github.com/kubernetes/kubernetes/pull/93405, we can't afford to introduce new flakes with CI in its current state

**Pull request merged**